### PR TITLE
UI - Bugfix: kv /edit/ and /create/ URLs

### PR DESCRIPTION
**Pull request by johncowen**
This is related to #4324 and the 301 redirects and somewhat similar to #4373 in that it involves trailing slashes.

Overall the problem relates to the fact that we use `/ui/kv/key-name/edit` and `/ui/kv/folder-name/create` 'endpoints' in the UI, and potentially these could be actual key/folder names (folder/in/here/create=value for example)

This is more or less dealt with when developing in the ember dev server as we don't have the additional complication of the 301 trailing slash redirects, whereas when things are baked into the consul binary the 301 removal of trailing slashes complicates matters.

Lastly, when testing manually (in the consul binary and in embers dev mode) everything works well, but when testing using embers testem test environment, I get a TransitionError so I'm not 100% happy to merge this myself yet. Mainly PR'ing so this doesn't get lost in the noise and to remind myself to get some more eyes on it to see if I can figure out where the TransitionError is coming from.

### CODE REVIEW

File: ui-v2/app/routes/dc/kv/create.js [hunk 1]
```diff
@@ -20,7 +24,14 @@ export default Route.extend(WithKvActions, {
       create: true,
       isLoading: false,
       item: this.item,
-      parent: repo.findBySlug(key, dc),
+      parent: repo.findBySlug(key, dc).catch(e => {
+        if (e.errors && e.errors[0] && e.errors[0].status == '404') {
+          const url = get(transition, 'intent.url');
+          if (url.endsWith('/create')) {
+            return this.transitionTo('dc.kv.folder', key + '/create/');
```
Comment (meirish) [inline ui-v2/app/routes/dc/kv/create.js]: This one is `transitionTo` instead of `replaceWith` - was there a reason for the difference?

File: ui-v2/app/routes/dc/kv/create.js [hunk 2]
```diff
@@ -3,14 +3,18 @@ import { inject as service } from '@ember/service';
...
 export default Route.extend(WithKvActions, {
   templateName: 'dc/kv/edit',
   repo: service('kv'),
-  beforeModel: function() {
+  beforeModel: function(transition) {
+    const url = get(transition, 'intent.url');
+    const search = '/create/';
+    if (url && url.endsWith(search)) {
+      return this.replaceWith('dc.kv.folder', this.paramsFor(this.routeName).key + search);
+    }
```
Comment (meirish) [inline ui-v2/app/routes/dc/kv/create.js]: This is "fun" to reason about 😬 - so if it ends in a slash, there's no way someone navigated there from refresh because the binary would cancel it. This is handling the case where they have a key named `create` then?

File: ui-v2/app/routes/dc/kv/create.js [hunk 3]
```diff
@@ -20,7 +24,14 @@ export default Route.extend(WithKvActions, {
...
       item: this.item,
-      parent: repo.findBySlug(key, dc),
+      parent: repo.findBySlug(key, dc).catch(e => {
+        if (e.errors && e.errors[0] && e.errors[0].status == '404') {
+          const url = get(transition, 'intent.url');
+          if (url.endsWith('/create')) {
+            return this.transitionTo('dc.kv.folder', key + '/create/');
+          }
+        }
+      }),
```
Comment (meirish) [inline ui-v2/app/routes/dc/kv/create.js]: And this is handling the case where they land here via the url being stripped of the trailing slash - do the data fetch, if it 404s, try it as a folder. I guess this could be where users land if something is deleted between fetching a list, and navigating to the detail as well.

File: ui-v2/tests/acceptance/dc/kvs/edit-create-keynames.feature [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      - [Action]
+    ---
+    When I visit the kvs page for yaml
+    ---
+      dc: datacenter
+      kv: foo/bar/[Action]/
+    ---
+    Then the url should be /datacenter/kv/foo/bar/[Action]/
+    And I see 1 kv model
+    # And the last GET request was made to "/v1/kv/foo/bar/[Action]/?keys&dc=datacenter&separator=%2F"
```
Comment (meirish) [inline ui-v2/tests/acceptance/dc/kvs/edit-create-keynames.feature]: heh random, but seeing this I was curious if you change the way keys / paths are rendered if they use a different separator, or if that's passed in and used when you parse keys

**Comment by johncowen** [comment]
Hey @meirish ,

Just so you know I'm getting to this, I'd really like to try and get to the bottom of why I get this TransitionAborted thing only in tests. I won't be merging this until I come back and have another crack and respond to your comments also. Should be over the next day or so.

Thanks

**Comment by johncowen** [comment]
Hey @meirish ,

Sorry about the delay. I had another crack and think about this yesterday.

I'm going to reply inline to some or your comments above in a sec. But just to give you the lowdown here.

So, looks like the `return this.replaceWith` in the `Promise.catch` was the gotcha for the ember tests. Worked fine when using the app, but when running in `testem` it would produce that `TransitionAborted` error. When I removed the `return` in the `catch` (so essentially return `null`), the tests passed and things continued to work when using the app.

Problem solved? Unfortunately not! :joy:

I'm still going to leave this unmerged for now as I'm still not entirely happy with it, and I've also noticed one thing that is 'unfixable' with the current approach.

Say I have a 'folder' in the root called `create/` with some further values inside. Within the `consul` ui slash stripping http server, if I directly visit the url to view this page:

`ui/kv/create/`

..then the trailing slash will be stripped off, making:

`ui/kv/create`

..which is the endpoint for creating a new kv within root - ouch! This means I shouldn't redirect back.

I've had a brief look at the old UI and I don't think this was ever an issue. Partly due to the fact that, as we were using the old `#/kv/` style of URLs, the `consul` trailing slash stripping wasn't an issue. Plus as we used to have the 'creation' page as part of the listing view (we had the 2 pane layout) a `/kv/create` URL didn't exist, you'd just create from the listing page at `/kv`.

I really wanted to keep the exact same URL structure when rebuilding the old UI, and I'd still really like to keep it exactly the same, but the only real way I can think of to maintain this would be to put the 'action verb' at the start of the URL. Alternatively we remove the slash stripping in the backend, which I don't think is the best either.

We are also talking about reintroducing the 2 pane thing, I'm pretty sure this won't help matters here as we will still maintain a 'list only' view and a 'list with create' view (each with its own URL), which means it will still be different from the old implementation.

Anyway, sorry ^ is a bit wordy/ramble-y, thought I'd brain dump a little here so you know what I'm thinking, might help me when I come back here also.

Thanks,

File: ui-v2/app/routes/dc/kv/create.js (hunk 1 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/kv/create.js]: Think I just missed this, have updated to use `replaceWith` everywhere

File: ui-v2/app/routes/dc/kv/create.js (hunk 2 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/kv/create.js]: Hehe yeah, it's certainly full of little edge cases.

File: ui-v2/app/routes/dc/kv/create.js (hunk 3 above)
Comment (johncowen) [reply ui-v2/app/routes/dc/kv/create.js]: Yup, so try and get a `create` key from the API, if its not there, the user could have meant `create/` but it was stripped. So try that by forwarding it back using `ember`.

> I guess this could be where users land if something is deleted between fetching a list, and navigating to the detail as well.

Yeah, if that happens, this is just the same as a 404, so like typing any old thing into the URL under `kv`. It will redirect you back up to the top of the `kv` tree.

File: ui-v2/tests/acceptance/dc/kvs/edit-create-keynames.feature (hunk 4 above)
Comment (johncowen) [reply ui-v2/tests/acceptance/dc/kvs/edit-create-keynames.feature]: Not sure I follow this. You mean the `separator=` query param yeah?

I know I'm always using the `separator=%2F` query param. There's no way via the UI to alter that right now (although it's a good candidate for a UI setting if there's a use case for that). So it always uses `/`.

Related to this, if we URL encoded the entire KV path in the URL that would solve all this 😁 . But then we lose the backwards compat of the URLs. Maybe that's a solution here? Would mean a bigger chat with more folks though.

**Comment by johncowen** [comment]
Yup, still thinking about this one 😂 

Had an idea, I reckon I could use query strings here to figure out the 'action'. It's a bit old skool but it would work here and I think it would be worth it, so:

`/ui/kv/key-name?edit`
`/ui/kv/edit?edit`

`/ui/kv/folder-name?create`
`/ui/kv/create?create`

ember probably won't like it, and it might mean doing some things that aren't 'strictly ballroom'. Will also mean a 'breaking' URL structure :(

I think the only alternative is going to be something like:

`/ui/edit-kv/key-name`
`/ui/edit-kv/edit`

`/ui/create-kv/folder-name`
`/ui/create-kv/create`

..which I don't particularly like.

Might give it a go after my current work. @meirish what do you think? Can you think of any gotchas?

**Comment by johncowen** [comment]
I'm going to close this, I think moving forwards I'll eventually do something with query params as mentioned here:

https://github.com/hashicorp/consul/pull/4411#issuecomment-426566777
